common/src/main/java/org/apache/omid/NetworkUtils.java

```diff
@@ -34,6 +36,14 @@
 
     public static String getDefaultNetworkInterface() {
 
+        try (DatagramSocket s=new DatagramSocket()) {
```

Comment:
This code in particular is meant to change behaviour.
It is actually part of OMID-157, to get the Travis test running, but the fix should apply to a lot of (virtualised) environments.

Reply:
Making sure I get it...

* Binds a socket on a random local port (ideally 127.0.0.1 for ipv4)
* This creates an InetAddress to 1.1.1.1
* Tries to bind to 1.1.1.1 on port 0 (what does that actually do?)
* Pulls the network interface used to do that bind, which is what we actually want

Any downside on this? Does it require being connected to the internet to work?

Reply:
Is there a reason we can't use `InetAddress.getLocalHost()` instead of something on the public internet?

Reply:
We specifically don't want to bind to localhost, we want to find to the "public" network interface that the clients can access.

The idea is that a random public IP address will be routed through the default route, which goes through the "public" network interface, and the clients will be able to access the server on the IP address of this "public' interface.

This is all just hacky guesswork as there is no way to definitely know which interface we want. What we really should do is make the public address configurable.

As for the actual implementation:

- The DatagramSocket() constructor will bind to 0.0.0.0, the wildcard address. (though we don't really care about the listening part)
- connect() will change the internal state of the Socket to only allow traffic to/from that address. We don't really care about that, what we do care about is that as a side effect, it will set the socket's localAddress to that of the outgoing interface that 1.1.1.1 is accessible via, and which we hope is the "public interface"
- After this, we will close the socket without having transmitted or received a single packet on it.

Revisiting this, we'd probably have a better chance of finding the right interface if we used an address from the ZK ensemble instead of 1.1.1.1, in case there's some weird networking setup in the cluster.

Furthermore, I think that the method of trying to figure out the interface, and then looking up the address of that interface is also flawed, as interfaces may have multiple addressed bound to them. A better approach would be using s.getLocalAddress() directly. 

Completely rewriting that logic is outside the scope of this ticket, though.

Reply:
Reading the code, the interface name IS configurable, which should cover 99% of real-world use cases.

Reply:
Ok. You've put a lot of thought into this :). Don't need to get hung up here. Thanks for your thoroughness, as usual.